lib/AST/DiagnosticConsumer.cpp

```diff
@@ -14,15 +14,148 @@
 //
 //===----------------------------------------------------------------------===//
 
-#define DEBUG_TYPE "swift-basic"
+#define DEBUG_TYPE "swift-ast"
 #include "swift/AST/DiagnosticConsumer.h"
 #include "swift/AST/DiagnosticEngine.h"
+#include "swift/Basic/Defer.h"
+#include "swift/Basic/SourceManager.h"
 #include "llvm/ADT/StringRef.h"
+#include "llvm/ADT/StringSet.h"
 #include "llvm/Support/Debug.h"
 #include "llvm/Support/raw_ostream.h"
 using namespace swift;
 
-DiagnosticConsumer::~DiagnosticConsumer() { }
+DiagnosticConsumer::~DiagnosticConsumer() = default;
+
+llvm::SMLoc DiagnosticConsumer::getRawLoc(SourceLoc loc) {
+  return loc.Value;
+}
+
+FileSpecificDiagnosticConsumer::FileSpecificDiagnosticConsumer(
+    SmallVectorImpl<ConsumerPair> &consumers)
+  : SubConsumers(std::move(consumers)) {
+#ifndef NDEBUG
+  llvm::StringSet<> seenFiles;
+  for (const ConsumerPair &consumerPair : SubConsumers) {
+    if (consumerPair.first.empty())
+      continue;
+    bool isUnique = seenFiles.insert(consumerPair.first).second;
+    assert(isUnique &&
```

Comment:
Why would duplicates be an error instead of just something to ignore?

Reply:
Because the way things are set up only one of the two consumers will get the diagnostics from that file, rather than both of them, and I don't think it's even guaranteed which (because of `std::sort`). We could change that if it ever mattered, but it's slightly harder to implement and we don't need it today.

Reply:
Add what you wrote above as a comment for the assert?